### DISCUSSION TITLE
Using dxWrapper for and reversing Gangsters: Organized Crime

### DISCUSSION THREAD

**Discussion by luckydom** [elishacloud/dxwrapper#208] [General]
Hi

I've been trying to get GOC working in windowed mode. My goal is to get it run in windowed mode before continuing writing other hooks. I know its possible to do it with dxwnd but the graphics become so awful and it is very laggy. So I tried overriding dwStyle that CreateWindowEx() function takes in. This seem to work for other games but not this one. I simply changed memory address to change the param for that function on the fly. I could see some resemblance of a window with title bar appear but it's completely unusable - for all intents and purposes, it still remained full screen. I think it has to do with compatibility layer for this game and DirectDraw in general. That's why I thought of completely migrating off of it and perhaps porting it to  SDL2. Has anyone done anything similar? Is it even possible? Or am I signing up for a sisyphean task? Any tips, links, opinions and other thoughts are welcome.

![image](https://github.com/elishacloud/dxwrapper/assets/26175304/9424261d-45a2-4617-833a-db4e35afe7d3)
![image](https://github.com/elishacloud/dxwrapper/assets/26175304/52c663fa-832f-4f66-b89c-cb0a80a5b984)
 
Thanks!

**Comment by elishacloud** [comment]
I am not sure of anyway to get Gangsters working in windowed mode.  The issue is that the game uses windows/GDI on top of the game graphics for starting menu.  There would have to be code to dynamically adjust all these graphics.  You could try [cnc-ddraw](https://github.com/FunkyFr3sh/cnc-ddraw).

**Reply by luckydom** [reply]
So that is my overall goal to write hook functions within a DLL that would dynamically adjust what's needed. I'm starting to explore what is needed, what APIs are being used, etc. So, makes me very interested in how dxWrapper replace some of these DirectDraw API calls. Tried CNC version, btw, - didn't work.

As for GDI, would replacing it with SDL2 work?

**Comment by elishacloud** [comment]
BTW: the dxwrapper project has ways to convert games to windowed mode, but it is only best-effort support and is not the focus of this project.  dxwrapper is more about making the games work as initially intended by the author, but on newer or more modern systems.

**Reply by luckydom** [reply]
Any docs on how can I try to run it in windowed mode using dxwrapper? Curious to at least give it a shot!

**Comment by elishacloud** [comment]
Here is the latest build, with the correct options enabled already:  [dxwrapper.zip](https://github.com/elishacloud/dxwrapper/files/11961002/dxwrapper.zip)

Here are the options needed for this:

```ini
[Compatibility]
Dd7to9                     = 1

[d3d9]
EnableWindowMode           = 1
WindowModeBorder           = 1
```

**Reply by luckydom** [reply]
Wow - definitely an improvement: very smooth pre-game screen and it is movable. I guess there is nothing to resize for so it's not needed.

However, for game screen, it's not movable, just sits in the center. Also, double mouse appears after alt+tabing since mouse is contained within the window and you'd like to access other programs on your screen. But it's not laggy! Mouse doesn't drag.

This is spectacular. What was changed and how?

**Comment by luckydom** [comment]
So I reinstalled the game and tried without any patches to make a windowed mode (overriding memory address on the fly) and somewhat succeeded but it's not a flexible windowed mode...  Problems are:
1. In-game it's quite laggy and window is immovable & not resizable so really it's full screen with a titlebar... 
2. Pre-game screen is movable but not resizable, also labels not visible but not too concerned about that given it's been solved before
![image](https://github.com/elishacloud/dxwrapper/assets/26175304/d0810cc3-cd68-4668-9b13-ab9f749e9875)
![image](https://github.com/elishacloud/dxwrapper/assets/26175304/d3c84457-1ab7-4c8e-8105-b22cef50b125)

**Comment by elishacloud** [comment]
> This is spectacular. What was changed and how?

I did not change anything, just gave you the latest build with the settings in the ini file set as seen above.

> 1. In-game it's quite laggy and window is immovable & not resizable

I am not seeing any lag in the game for me.  It could be something with the build I gave you.  But the window being unmovable is something to do with how the game takes control of the cursor when you click on the window.  I'm sure it is fixable, but not something I will look into soon.

> Pre-game screen is movable but not resizable

Yes, the window will not be resizable.  That is expected.  There is no code in dxwrapper to allow statically sized windows to be resizable.  Anyways, it would be difficult with this game to allow resizing the window without getting parts of the graphics out of alignment.  The game is using a combination of DirectDraw and GDI to draw the screen and scaling all the GDI parts to match the new window size would not be easy.
